### PR TITLE
Instance Norm.

### PR DESCRIPTION
**Pull request by iamshnoo**
Instance Norm Layer.

***Note :*** Currently "InstanceNormLayerTest" and "InstanceNormLayerParametersTest" pass. "GradientInstanceNormLayerTest" crashes and gives weird memory errors for reasons unknown to me as of yet (I need some help debugging this, please.) I am probably missing something minor with the Gradient test calculations.

To compare armadillo functionality implementation with PyTorch implementations, take a look at this [Google Colab notebook](https://colab.research.google.com/drive/18E8a-LTHFOpqa6w53vOsjRz7nbRHLvaf?usp=sharing).

The basic idea followed for this implementation is that, Instance Norm applied on a (N,C,H,W) input is equivalent to Batch Norm applied on a (1,N*C,H,W) input. So, this layer is implemented as a BatchNorm wrapper layer.
This idea is same as the one followed by the original author of the paper [here](https://github.com/DmitryUlyanov/texture_nets/blob/master/InstanceNormalization.lua) and also the same as the PyTorch implementation [here](https://github.com/pytorch/pytorch/pull/1283).

### CODE REVIEW

File: src/mlpack/methods/ann/layer/instance_norm.hpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+   * @param size The number of input units / channels.
+   * @param eps The epsilon added to variance to ensure numerical stability.
+   * @param average Boolean to determine whether cumulative average is used for
+   *                updating the parameters or momentum is used.
+   * @param momentum Parameter used to to update the running mean and variance.
+   */
+  InstanceNorm(const size_t size,
+              const double eps = 1e-5,
+              const bool average = true,
+              const double momentum = 0.1);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/instance_norm.hpp:72]:
```suggestion
               const double eps = 1e-5,
               const bool average = true,
               const double momentum = 0.1);
```

Just to align the parameter with the first one.

File: src/mlpack/methods/ann/layer/instance_norm_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    shapeB = input.n_cols;
+    batchNorm = ann::BatchNorm<> (size*input.n_cols, eps, average, momentum);
+    batchNorm.Reset();
+    runningMean.zeros(size, 1);
+    runningVariance.ones(size, 1);
+    runningVariance = batchNorm.TrainingVariance();
+    reset = true;
+  }
+
+  if(deterministic)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/instance_norm_impl.hpp:71]:
```suggestion
  if (deterministic)
```

Minor style issue, missing space.

**Comment by zoq** [comment]
Not sure, but it looks like if we resolve the merge conflict the memory check will pass.

**Comment by iamshnoo** [comment]
Keep open please, thanks!

**Comment by hello-fri-end** [comment]
@zoq @iamshnoo : Hi! I've figured out why "GradientInstanceNormLayersTest" crashes. The batchNorm object is initialized in the Forward() function of InstanceNorm layer, so when ResetParameters() is called by FFN, memory isn't allotted for the parameters ( & subsequently gradient) of InstanceNorm layer. Hence, the memory crash.

We could initialize the batchNorm object in the constructor itself. Although, we would need to add input.n_cols to it's aruguments.

Please let me know what you think about this.

**Comment by iamshnoo** [comment]
@hello-fri-end yep, that seems like a possible solution. I am concerned about having to pass in ```input.n_cols``` as a parameter to the class however. Ideally, we should only be having the input parameters that are currently there. But, unless there is any other way around it, I guess your idea is the only way to move forward with this layer.

Feel free to create a new PR with the commits in this branch, and add the modifications. If the tests pass in that, and nobody has any issues with the addition of the extra parameter, that's good enough for me.
